examples/pixelcnn/train.py

```diff
@@ -210,32 +186,46 @@ def train(pcnn_module, model_dir, batch_size, init_batch_size, num_epochs,
   eval_iter = iter(eval_ds)
 
   # Compute steps per epoch and nb of eval steps
-  steps_per_epoch = data_source.TRAIN_IMAGES // batch_size
-  steps_per_eval = data_source.EVAL_IMAGES // batch_size
+  if config.num_train_steps == -1:
+    steps_per_epoch = (
+      data_source.info.splits['train'].num_examples // config.batch_size
+    )
+  else:
+    steps_per_epoch = config.num_train_steps
```

Comment:
So it is not possible to have `num_train_steps > num_train_examples` ?
I would have assumed that you can specify e.g. `num_train_steps = 2 * num_train_examples` which would result in the same as `num_epochs = 2`.

(I think that's also what I have seen in similar configurations in other projects, though AFAIK we don't use this in Flax examples yet.)

Reply:
Hmm TBH I actually didn't think of this scenario at all.

Some thoughts:

by `num_train_examples` do you mean `data_source.info.splits['train'].num_examples`? So even with mock data and explicitly specifying `num_examples`, `data_source.info.splits['train'].num_examples` would still return the number of examples in the split information from original metadata i.e 50000 for train split. See warning here: https://www.tensorflow.org/datasets/api_docs/python/tfds/testing/mock_data. 

Also since we do `ds.repeat()` in the input_pipeline, so we can't really do `len(list(ds))`. 

So IMO this shouldn't really effect the training loop?

Reply:
Yes, 'data_source.info.splits['train'].num_examples == num_train_examples` (please forgive my laziness :-)

IMO it would be great to have `config.num_train_steps` for total number of training step. This makes it easy to set a target for total number of examples. Alternatively, if `config.num_train_steps == -1`, it would use `num_train_steps = config.num_epochs * steps_per_epoch`. This would allow flexibly to define either in the config, and would also make it possible to stop after a single step in the config regardless of the value of `data_source.info.splits['train'].num_examples`.

(Otherwise, if you prefer keeping the current implementation, I would suggest to rename to `config.num_steps_per_epoch` - also, if it's only used for testing, then you might even use a single config param for this.)

Reply:
Cool, thanks for the suggestion! Made the necessary changes.